### PR TITLE
Change Variants to be a ShortVec<Variant>

### PR DESCRIPTION
**Pull request by yzhang1994**
PR 1/2 for issue #1800

### CODE REVIEW

File: components/locid/src/subtags/variants.rs [hunk 1]
```diff
@@ -127,7 +123,7 @@ impl Variants {
     ///
     /// This function accepts any [`Box`]`<`[`Variant`]`>` that is expected to be a
     /// valid list of sorted variants.
-    pub const unsafe fn from_raw_unchecked(input: Option<Box<[Variant]>>) -> Self {
+    pub const unsafe fn from_raw_unchecked(input: ShortVec<Variant>) -> Self {
```
Comment (sffc) [inline components/locid/src/subtags/variants.rs:126]: @zbraniecki These `raw` methods are unused in ICU4X outside tests. Do you have an objection to removing them? Doing so would bring the following benefits:

1. We could keep `ShortVec` private more easily
2. It's weird to have unchecked methods that are public and safe

If you want to keep them, do you have a suggestion on what signature to use such that we can avoid making `ShortVec` be public?

**Review by sffc — CHANGES_REQUESTED** [review]
I'd rather not make `ShortVec` be public. I requested feedback from Zibi.

**Comment by zbraniecki** [comment]
So, this will change Variants from `Box<[Variant]>` to `Vec<Variant>`. Is that what we want?

**Comment by sffc** [comment]
> So, this will change Variants from `Box<[Variant]>` to `Vec<Variant>`. Is that what we want?

It changes to `ShortVec<Variant>`, not `Vec<Variant>`. `ShortVec` is optimized for vectors of length 0 and 1.

This is what we want because it enables us to const-construct more types of locales without needing a heap allocation.

**Comment by zbraniecki** [comment]
`SortVec` is an enum that has `Vec`. Maybe what I'm asking is if we should consider having `ShortStaticVec` that has `Empty, One, Box<[]>` ?

**Comment by sffc** [comment]
> `SortVec` is an enum that has `Vec`. Maybe what I'm asking is if we should consider having `ShortStaticVec` that has `Empty, One, Box<[]>` ?

This is internal, so we can switch in the future if a need arises.

That said, I recall similar questions coming up before, and my recollection of the conclusion was that we could not identify a concrete reason to favor `Box<[T]>` over `Vec<T>`. If there is evidence to the contrary, could you share it?

In any case, I hope that this can be in a follow-up.

File: components/locid/src/subtags/variants.rs (hunk 1 above)
Comment (sffc) [reply components/locid/src/subtags/variants.rs:126]: @yzhang1994 Please remove the two methods that use ShortVec in their public signature and then move ShortVec back to being crate-private.

Comment (yzhang1994) [reply components/locid/src/subtags/variants.rs]: Done. Remove both methods & made ShortVec crate-private.

**Review by sffc — APPROVED** [review]

**Review by Manishearth — APPROVED** [review]

**Comment by hsivonen** [comment]
Looks like CI has been failing since this PR landed.

**Comment by hsivonen** [comment]
The merge annotation here says "27 checks passed", but https://github.com/unicode-org/icu4x/commits/main shows a red x next to the changeset landed here.

**Comment by zbraniecki** [comment]
> Now, do we want to reduce the size of it though? Over 200 bytes is a lot for something that is usually less than 10 bytes of actual data.

I'd love to try to fix it. I think it's the hottest item in the whole ICU4X.

**Comment by zbraniecki** [comment]
can we consider reverting this as well?

**Comment by sffc** [comment]
> can we consider reverting this as well?

We haven't discussed stack size of locale as being an issue before. This PR changed the stack size from 184 (already quite big) to 216, which is only a 17% increase. Reverting this PR will not fix the underlying problem, which is a deeply nested struct layout that probably results in a lot of wasted bytes.

If we want to start measuring the stack size of Locale, and in general of other ICU4X types, then that should be the subject of another issue.

FYI, the purpose of this PR is to make more things in Locale be const-constructible, which is an impactful issue to solve because we can start writing `locale!("ca-ES-valencia")` and eventually `locale!("en-u-ca-hebrew")`. This benefits developer experience and i18n correctness since it encourages the use of extension keywords.

We're already optimizing for a lot of things (performance, code size, heap memory, developer experience, i18n correctness), so before we add another thing to that list, we should make sure we can justify why it is an impactful thing to measure.

**Comment by sffc** [comment]
FYI, the large majority of the stack size of Locale is `extensions::Extensions`. I added more size tests to demonstrate: https://github.com/unicode-org/icu4x/pull/2078

It's true that this PR likely increases the stack size of `Variants`, which likely impacts `LanguageIdentifier` to a greater percentage degree than `Locale`. I may do a quick check to see if there's a way to mitigate that cost in `ShortVec`.

**Comment by sffc** [comment]
A quick investigation suggests that using `Box<[T]>` instead of `Vec<T>`, as @zbraniecki suggested, positively improves stack size (presumably by removing the capacity slot). We can get `Locale` down from 216 to 200 by making this change, or 50% of the regression introduced by using ShortVec.

**Comment by sffc** [comment]
I filed an issue explaining how we can *fully eliminate* the stack size regression without sacrificing functionality.

https://github.com/unicode-org/icu4x/issues/2084

**Pull request merged**